File: android/app/src/main/kotlin/com/brianhenning/cribbage/HumanManager.kt
package com.brianhenning.cribbage

import android.app.Application
import com.humansecurity.mobile_sdk.HumanSecurity
import com.humansecurity.mobile_sdk.main.HSBotDefenderChallengeResult
import com.humansecurity.mobile_sdk.main.policy.HSAutomaticInterceptorType
import com.humansecurity.mobile_sdk.main.policy.HSPolicy
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

class HumanManager {
    companion object {
        fun start(application: Application) {
            try {
                val policy = HSPolicy()
                policy.automaticInterceptorPolicy.interceptorType =
                    HSAutomaticInterceptorType.NONE
                HumanSecurity.start(application, "PX_APP_ID", policy)
            } catch (exception: Exception) {
                println("Exception: ${exception.message}")
            }
        }

        fun handleEvent(
            call: MethodCall,
            result: MethodChannel.Result
        ) {
            if (call.method == "humanGetHeaders") {
                val json = JSONObject(
                    HumanSecurity.BD.headersForURLRequest(null)
                        as Map<*, *>
                )
                result.success(json.toString())
            } else if (call.method == "humanHandleResponse") {
                val handled = HumanSecurity.BD.handleResponse(
                    call.arguments!! as String
                ) { challengeResult: HSBotDefenderChallengeResult ->
                    result.success(
                        if (challengeResult ==
                            HSBotDefenderChallengeResult.SOLVED
                        ) "solved" else "cancelled"
                    )
                    null
                }
                if (!handled) {
                    result.success("false")
                }
            } else {
                result.notImplemented()
            }
        }
    }
}

File: ios/Runner/HumanManager.swift
import Foundation
import Flutter
import HUMAN

class HumanManager {
    static let shared = HumanManager()

    func start() {
        do {
            let policy = HSPolicy()
            policy.automaticInterceptorPolicy.interceptorType = .none
            try HumanSecurity.start(appId: "PX_APP_ID", policy: policy)
        } catch {
            print("Error: \(error)")
        }
    }

    func setupChannel(with controller: FlutterViewController) {
        let humanChannel = FlutterMethodChannel(
            name: "com.humansecurity/sdk",
            binaryMessenger: controller.binaryMessenger
        )
        humanChannel.setMethodCallHandler {
            (call: FlutterMethodCall, result: @escaping FlutterResult) in

            if call.method == "humanGetHeaders" {
                var json: String?
                do {
                    let headers = HumanSecurity.BD
                        .headersForURLRequest(forAppId: "PX_APP_ID")
                    let data = try JSONSerialization
                        .data(withJSONObject: headers)
                    json = String(data: data, encoding: .utf8)
                } catch {
                    print("Error: \(error)")
                }
                result(json)
            } else if call.method == "humanHandleResponse" {
                if let response = call.arguments as? String,
                   let data = response.data(using: .utf8),
                   let httpURLResponse = HTTPURLResponse(
                       url: URL(string: "https://example.com")!,
                       statusCode: 403,
                       httpVersion: nil,
                       headerFields: nil
                   ) {
                    let handled = HumanSecurity.BD.handleResponse(
                        response: httpURLResponse,
                        data: data
                    ) { challengeResult in
                        result(
                            challengeResult == .solved
                                ? "solved" : "cancelled"
                        )
                    }
                    if handled {
                        return
                    }
                }
                result("false")
            } else {
                result("")
            }
        }
    }
}
